### PR TITLE
Element availability timeout implemented (close #458)

### LINKED ISSUE #458: Implement action target waiting timeout option

### PR DESCRIPTION
**Pull request by helen-dikareva**
/cc @AlexanderMoskovkin @AndreyBelym

### CODE REVIEW

File: src/legacy/test-run/index.js [hunk 1]
```diff
@@ -49,26 +50,30 @@ export default class LegacyTestRun extends Session {
...
         var payloadScript = Mustache.render(IFRAME_TEST_RUN_TEMPLATE, {
-            sharedJs:               sharedJs,
-            takeScreenshotsOnFails: this.opts.takeScreenshotsOnFails,
-            skipJsErrors:           this.opts.skipJsErrors,
-            nativeDialogsInfo:      JSON.stringify(this.nativeDialogsInfo)
+            sharedJs:                   sharedJs,
+            takeScreenshotsOnFails:     this.opts.takeScreenshotsOnFails,
+            skipJsErrors:               this.opts.skipJsErrors,
+            nativeDialogsInfo:          JSON.stringify(this.nativeDialogsInfo),
+            elementAvailabilityTimeout: this.opts.elementAvailabilityTimeout === void 0 ?
```
Comment (inikulin) [inline src/legacy/test-run/index.js]: Why we need it here? We already did it in `Runner`.

File: test/functional/fixtures/waiting-element-timeout/es-next/test.js [hunk 2]
```diff
@@ -0,0 +1,16 @@
+var expect = require('chai').expect;
+
+describe('[API] Waiting element timeout', function () {
+    it('Should pass if waiting element timeout more then time of element appearance', function () {
+        return runTests('testcafe-fixtures/waiting-element-timeout-test.js', 'Wait for element appearance before click',
+            { elementAvailabilityTimeout: 2000 });
```
Comment (inikulin) [inline test/functional/fixtures/waiting-element-timeout/es-next/test.js]: Let's just set it for all tests inside `runTests`

File: test/functional/fixtures/waiting-element-timeout/es-next/test.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+var expect = require('chai').expect;
+
+describe('[API] Waiting element timeout', function () {
```
Comment (inikulin) [inline test/functional/fixtures/waiting-element-timeout/es-next/test.js]: It's not part of test API and it should be in the `fixtures` dir

File: test/functional/setup.js [hunk 4]
```diff
@@ -1,150 +1,155 @@
...
+        .then(function () {
+            global.testReport = null;
+
+            global.runTests = function (fixture, testName, opts) {
+                var report                = '';
+                var runner                = testCafe.createRunner();
+                var fixturePath           = path.join(path.dirname(caller()), fixture);
+                var skipJsErrors          = opts && opts.skipJsErrors;
+                var quarantineMode        = opts && opts.quarantineMode;
+                var elementAvailabilityTimeout = opts && opts.elementAvailabilityTimeout;
```
Comment (inikulin) [inline test/functional/setup.js]: Can we set it to `200` by default for all tests except those which specify timeout explicitly?

**Comment by inikulin** [comment]
\r-

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
FPR

File: test/functional/fixtures/waiting-element-timeout/pages/index.html [hunk 5]
```diff
@@ -0,0 +1,18 @@
...
+</head>
+<body>
+<button id="button1">Start of element appearance</button>
+<button id="button2" style="display: none">After 1500 ms</button>
+
+<script>
+    document.getElementById('button1').addEventListener('click', function () {
+        window.setTimeout(function () {
+            document.getElementById('button2').style.display = "block";
+        }, 2000);
```
Comment (AlexanderMoskovkin) [inline test/functional/fixtures/waiting-element-timeout/pages/index.html]: it should be `1500` here

File: src/runner/index.js [hunk 6]
```diff
@@ -121,9 +125,11 @@ export default class Runner extends EventEmitter {
         return this;
     }
 
-    run ({ skipJsErrors, quarantineMode } = {}) {
-        this.opts.skipJsErrors   = !!skipJsErrors;
-        this.opts.quarantineMode = !!quarantineMode;
+    run ({ skipJsErrors, quarantineMode, elementAvailabilityTimeout } = {}) {
+        this.opts.skipJsErrors               = !!skipJsErrors;
+        this.opts.quarantineMode             = !!quarantineMode;
+        this.opts.elementAvailabilityTimeout = elementAvailabilityTimeout === void 0 ?
```
Comment (AlexanderMoskovkin) [inline src/runner/index.js]: It's a public API option and we can get here anything (function, object, number). What we should do here, what the best practice? Maybe check if `typeof elementAvailabilityTimeout !== 'number'` set it to default value? Or raise an exception if it's not a nubmer? /cc @inikulin

Comment (inikulin) [inline src/runner/index.js]: We don't do any checks for the programmatical API

**Comment by AlexanderMoskovkin** [comment]
ok

/r- just correct a wrong integer in the test and lgtm

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

File: test/functional/fixtures/api/raw/click/test.js [hunk 7]
```diff
@@ -48,7 +48,8 @@ describe('[Raw API] Click action', function () {
     });
 
     it('Should wait for the next action element to appear', function () {
-        return runTests('./testcafe-fixtures/click.testcafe', 'Wait for next action target appeared', { shouldFail: true })
+        return runTests('./testcafe-fixtures/click.testcafe', 'Wait for next action target appeared',
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/api/raw/click/test.js]: rename this test pls where it resides
```
        return runTests('./testcafe-fixtures/click.testcafe', 'Wait for the next action target to appear',
```

File: test/functional/fixtures/waiting-element-timeout/test.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
+var expect = require('chai').expect;
+
+describe('Waiting element timeout', function () {
+    it('Should pass if waiting element timeout more then time of element appearance', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/waiting-element-timeout/test.js]:
```
    it('Should pass if the element's availability timeout exceeds the waiting time for the element appearance', function () {
```

File: test/functional/fixtures/waiting-element-timeout/test.js [hunk 9]
```diff
@@ -0,0 +1,34 @@
...
+        return runTests('testcafe-fixtures/waiting-element-timeout-test.js', 'Wait for element appearance before click',
+            { elementAvailabilityTimeout: 2500 });
+    });
+
+    it('Should fail if waiting element timeout less then time of element appearance', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/waiting-element-timeout/test.js]:
```
    it('Should fail if the element's availability timeout is less than time required for the element to appear', function () {
```

File: test/functional/fixtures/waiting-element-timeout/test.js [hunk 10]
```diff
@@ -0,0 +1,34 @@
...
+        return runTests('testcafe-fixtures/waiting-element-timeout-test.js', 'Wait for element appearance before click',
+            { shouldFail: true, elementAvailabilityTimeout: 500 })
+            .catch(function (errs) {
+                expect(errs[0]).to.contains('The element that matches the specified selector is not visible.');
+            });
+    });
+
+    it('Should fail if waiting element timeout less then time of element appearance (legacy)', function () {
```
Comment (VasilyStrelyaev) [inline test/functional/fixtures/waiting-element-timeout/test.js]:
```
    it('Should fail if the element's availability timeout is less than the time required for the element to appear (legacy)', function () {
```

**Comment by VasilyStrelyaev** [comment]
\r-

**Comment by helen-dikareva** [comment]
This PR'll be merged after DragCommand https://github.com/DevExpress/testcafe/pull/456

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
FPR

**Comment by AndreyBelym** [comment]
lgtm

**Comment by VasilyStrelyaev** [comment]
lgtm

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by inikulin** [comment]
@helen-dikareva Tests fail due to linting error

**Comment by helen-dikareva** [comment]
Ooops, I didn't notice.Fixed now.

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
@testcafe-build-bot \retest

**Comment by helen-dikareva** [comment]
FPR

**Comment by AndreyBelym** [comment]
lgtm

**Comment by AlexanderMoskovkin** [comment]
lgtm

**Pull request merged**